locator/tablets.cc

```diff
@@ -435,6 +435,62 @@ load_stats& load_stats::operator+=(const load_stats& s) {
     return *this;
 }
 
+tablet_range_splitter::tablet_range_splitter(schema_ptr schema, const tablet_map& tablets, host_id host, const dht::partition_range_vector& ranges)
+    : _schema(std::move(schema))
+    , _ranges(ranges)
+    , _ranges_it(_ranges.begin())
+{
+    // Filter all tablets and save only those that have a replica on the specified host.
+    for (auto tid = std::optional(tablets.first_tablet()); tid; tid = tablets.next_tablet(*tid)) {
+        const auto& tablet_info = tablets.get_tablet_info(*tid);
+
+        auto replica_it = std::ranges::find_if(tablet_info.replicas, [&] (auto&& r) { return r.host == host; });
+        if (replica_it == tablet_info.replicas.end()) {
+            continue;
+        }
+
+        auto token_range = tablets.get_token_range(*tid);
+        auto range = dht::partition_range(
+                dht::partition_range::bound(dht::ring_position(token_range.start()->value(), dht::ring_position_view::token_bound::end), false),
+                dht::partition_range::bound(dht::ring_position(token_range.end()->value(), dht::ring_position_view::token_bound::end), true));
+        _tablet_ranges.emplace_back(range_split_result{replica_it->shard, std::move(range)});
+    }
+    _tablet_ranges_it = _tablet_ranges.begin();
+}
+
+std::optional<tablet_range_splitter::range_split_result> tablet_range_splitter::operator()() {
+    if (_ranges_it == _ranges.end() || _tablet_ranges_it == _tablet_ranges.end()) {
+        return {};
+    }
```

Comment:
This is an infinite interval, no? should return an empty interval.

Reply:
No, the splitter has traversed the ranges and there are no more intersections to return.
This is the same as `interval<>::intersect()`, returning disengaged optional.